File: Data Structures & Algorithms/binary-tree-diameter/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    diameterOfBinaryTree(root) {
        const result = [0];
        this.dfs(root, result);
        return result[0];
    }
    dfs(root, result) {
        if (root === null) {
            return 0;
        }
        const left = this.dfs(root.left, result);
        const right = this.dfs(root.right, result);
        result[0] = Math.max(result[0], left + right);
        return 1 + Math.max(left, right);
    }
}

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> res = new LinkedList();    
        dfs(res, root, 0);
        return res;
    }

    private void dfs(List<Integer> res, TreeNode node, int depth) {
        if (node == null) {
            return;
        }

        if (res.size() == depth) {
            res.add(node.val);
        }

        dfs(res, node.right, depth + 1);
        dfs(res, node.left, depth + 1);
    }
}
